modules/ROOT/pages/dataweave-type-system.adoc

```diff
@@ -0,0 +1,291 @@
+
+----
+fun findUserIndex(arr: Array<User>, condition: (userInArray: User, targetUser: User) -> Boolean): User = ???
+----
+
+If we call the function `findUserIndex` with some function that does not match the `condition` constraint `(User, User) -> Boolean`, DataWeave throws a type checking error.
+
+
+== Complex Types
+
+=== Any and Nothing
+
+In some cases you can not enforce a restriction as it accepts all the values. +
+
+* `Any` type accepts all possible values.
+* `Nothing` type accepts no value but it can be assigned to all the types. This type is not frequently used explicitly but it is by the type inference algorithm.
+
+=== Union Type
+
+Union type is used to compose types. The syntax to define a union type is:
+
+----
+TypeA | TypeB | ...
+----
+
+The following example defines a type that accepts any String and any Number:
+
+----
+var age: String | Number = ???
+----
+
+A common pattern is to use Null with a union type to specify that a type accepts null as a value. In the following example, the variable `a` accepts that the expression `payload.foo` returns `null`.
+
+----
+var a: String| Null = payload.foo
+----
+
+
+=== Intersection Type
+
+The intersection type intersects object types. In this case, the intersection works as the concatenation (`++`) of object types.
+
+_Introduced in DataWeave 2.3.0. Supported by Mule 4.3 and later._
+
+The syntax for intersection type is:
+----
+TypeA & TypeB & ...
+----
+
+In the following example, the intersection concatenates the two object types resulting in the type `{name: String, lastName: String}` This is an open object so it can have more key value pairs. The variable accepts the value assigned to it:
+
+----
+var a: {name: String} & {lastName: String} = {name: "John", lastName: "Smith", age: 34}
+----
+
+In the case of close objects, it returns the concatenation of the object types but resulting in a close object. +
+In the following example, the intersection results in the type `{|name: String, lastName: String|}`. This throws
+an exception because a close object does not accept an object value if there are additional fields in the object (field `age`):
+
+----
+var a: {|name: String|} & {|lastName: String|} = {name: "John", lastName: "Smith", age: 34}
+----
+
+
+=== Literal Types
+
+A literal type represents exactly one value. For example, the String value `"foo"` can be represented with the type `"foo"`.
+
+_Introduced in DataWeave 2.3.0. Supported by Mule 4.3 and later._
+
+The following literal types have been added to the type system:
+
+* String literal types
+* Number literal types
+* Boolean literal types
+
+You can use literal types with union types when you want to declare a type as a finite set of allowed values. For example, the following types declarations are aliases of union and literals types:
+
+----
+type Weekdays = "Saturday" | "Sunday"
+type Days = "Monday" | "Tuesday" | "Wednesday" | "Thursday" | "Friday" | "Saturday" | "Sunday"
```

Comment:
```suggestion
You can use literal types with union types when you want to declare a type as a finite set of allowed values. For example, the following types declarations are aliases for unions of literal types:

----
type Weekdays = "Monday" | "Tuesday" | "Wednesday" | "Thursday" | "Friday" 
type Days = Weekdays | "Saturday" | "Sunday"
```
Weekdays is used as type in the next type declaration (this is a good practice I think).